yjit/src/codegen.rs

```diff
@@ -5665,8 +5661,43 @@ fn gen_send_iseq(
         }
     }
 
-    // Number of locals that are not parameters
-    let num_locals = unsafe { get_iseq_body_local_table_size(iseq) as i32 } - (num_params as i32);
+    // Check if we need the arg0 splat handling of vm_callee_setup_block_arg
+    let arg_setup_block = captured_opnd.is_some(); // arg_setup_type: arg_setup_block (invokeblock)
+    let block_arg0_splat = arg_setup_block && argc == 1 && unsafe {
+         get_iseq_flags_has_lead(iseq) && !get_iseq_flags_ambiguous_param0(iseq)
+    };
+    if block_arg0_splat {
+        // If block_arg0_splat, we still need side exits after splat, but
+        // doing push_splat_args here disallows it. So bail out.
+        if flags & VM_CALL_ARGS_SPLAT != 0 && !iseq_has_rest {
+            gen_counter_incr!(asm, invokeblock_iseq_arg0_args_splat);
+            return CantCompile;
+        }
+        // The block_arg0_splat implementation is for the rb_simple_iseq_p case,
+        // but doing_kw_call means it's not a simple ISEQ.
+        if doing_kw_call {
+            gen_counter_incr!(asm, invokeblock_iseq_arg0_has_kw);
+            return CantCompile;
+        }
+    }
+    let splat_array_length = if flags & VM_CALL_ARGS_SPLAT != 0 && !iseq_has_rest {
+        let array = jit.peek_at_stack(ctx, if block_arg { 1 } else { 0 }) ;
+        let array_length = if array == Qnil {
+            0
+        } else {
+            unsafe { rb_yjit_array_len(array) as u32}
+        };
+
+        if opt_num == 0 && required_num != array_length as i32 + argc - 1 {
+            gen_counter_incr!(asm, send_iseq_splat_arity_error);
+            return CantCompile;
+        }
+        Some(array_length)
+    } else {
+        None
+    };
+
+    // We will not have CantCompile from here. You can use stack_pop / stack_pop.
```

Comment:
For a `CantCompile` exit to use the latest `ctx` (which will have the latest spill state), its `stack_size` should not be modified before returning `CantCompile`.

We're modifying `stack_size` in L5072-5717, so we shouldn't have `return CantCompile` after that. The code you highlighted was such `return CantCompile` instances after the `stack_pop`. We want to return `CantCompile` before that. Lazily doing the `stack_pop` seemed non-trivial because of those `CantCompile`, so I moved `CantCompile` instead.

Alternatively, we could rewind `Assembler` to the point before the YARV instruction is compiled, but in https://github.com/ruby/ruby/pull/7583, it felt like you didn't like that idea, so I took this path.